source/material_model/drucker_prager.cc

```diff
@@ -29,56 +31,134 @@ namespace aspect
+                                        :
+                                        cohesion * cos_phi + pressure * sin_phi );
+
+              // Rescale the viscosity back onto the yield surface
+              const double eta_plastic = strength * strain_rate_effective_inv;
+
+
+              const double effective_viscosity = 1.0 / ( ( 1.0 / ( eta_plastic + minimum_viscosity ) ) + ( 1.0 / maximum_viscosity ) );
+
+              Assert(dealii::numbers::is_finite(effective_viscosity),ExcMessage ("Error: Viscosity is not finite."));
+              if (derivatives != NULL)
+                {
+                  if (std::sqrt(strain_rate_effective) > std::numeric_limits<double>::min())
+                    {
+                      const double averaging_factor = effective_viscosity * effective_viscosity * std::pow(eta_plastic + minimum_viscosity,-2);
+                      effective_viscosity_strain_rate_derivatives = averaging_factor * 0.5 * (eta_plastic * (-1/std::sqrt(edot_ii_strict * edot_ii_strict))) * strain_rate_deviator;
+                      effective_viscosity_pressure_derivatives = averaging_factor * (dim == 2 ?
+                                                                                     sin_phi * strain_rate_effective_inv
+                                                                                     :
+                                                                                     (sin_phi*strength_inv_part));
+                    }
+                  else
+                    {
+                      effective_viscosity_strain_rate_derivatives = 0;
+                    }
+                }
+
+              if (std::sqrt(strain_rate_effective) <= std::numeric_limits<double>::min())
+                {
+                  out.viscosities[i] = maximum_viscosity;
+                }
+              else
+                {
+                  out.viscosities[i] = effective_viscosity;
+                }
+
+              Assert(dealii::numbers::is_finite(out.viscosities[i]),ExcMessage ("Error: Averaged viscosity is not finite."));
+
+              if (derivatives != NULL)
+                {
+
+                  derivatives->viscosity_derivative_wrt_strain_rate[i] = effective_viscosity_strain_rate_derivatives;
+
+                  derivatives->viscosity_derivative_wrt_pressure[i] = effective_viscosity_pressure_derivatives;
+
+                  Assert(dealii::numbers::is_finite(derivatives->viscosity_derivative_wrt_pressure[i]),
+                         ExcMessage ("Error: Averaged dviscosities_dpressure is not finite."));
+                  for (int x = 0; x < dim; x++)
+                    for (int y = 0; y < dim; y++)
+                      Assert(dealii::numbers::is_finite(derivatives->viscosity_derivative_wrt_strain_rate[i][x][y]),
+                             ExcMessage ("Error: Averaged dviscosities_dstrain_rate is not finite."));
+
```

Comment:
Why do you separate this if-block from the one above? If you could unify them it would be easier to read the code, because all derivative stuff would be in one place.